### PR TITLE
GUAC-1389: Add screen sharing support to VNC.

### PR DESCRIPTION
**Pull request by mike-jumper**

### CODE REVIEW

File: src/protocols/vnc/display.c [hunk 1]
```diff
@@ -0,0 +1,207 @@
...
+    free(buffer);
+
+}
+
+void guac_vnc_copyrect(rfbClient* client, int src_x, int src_y, int w, int h, int dest_x, int dest_y) {
+
+    guac_client* gc = rfbClientGetClientData(client, GUAC_VNC_CLIENT_KEY);
+    guac_vnc_client* vnc_client = (guac_vnc_client*) gc->data;
+
+    /* For now, only use default layer */
```
Comment (jmuehlner) [inline src/protocols/vnc/display.c]: For now?

File: src/protocols/vnc/display.c [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+
+    vnc_client->copy_rect_used = 1;
+
+}
+
+void guac_vnc_set_pixel_format(rfbClient* client, int color_depth) {
+    switch(color_depth) {
+        case 8:
+            client->format.depth        = 8;
+            client->format.bitsPerPixel = 8;
```
Comment (jmuehlner) [inline src/protocols/vnc/display.c]: Why use bitsPerPixel here, and bytes per pixel elsewhere?

File: src/protocols/vnc/display.c (hunk 1 above)
Comment (mike-jumper) [inline src/protocols/vnc/display.c]: This is an ancient comment, but I'm not sure what the plan was which would involve anything *but* default... I'll fix this.

File: src/protocols/vnc/display.c (hunk 2 above)
Comment (mike-jumper) [inline src/protocols/vnc/display.c]: Guacamole always represents color depth as bits per pixel. We only use bytes per pixel in the locations that libvncclient mandates it.

Looking into this revealed that the comment for `guac_vnc_set_pixel_format()` is wrong, though (in `display.h`). It's not a callback - it's a helper for updating the `rfbClient`.

Comment (mike-jumper) [inline src/protocols/vnc/display.c]: Also, the line in question is the `bitsPerPixel` member of a structure from libvncclient - we don't control that (though, if we did, the only difference would be lowercase and underscores).

File: src/protocols/vnc/settings.h [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+     * milliseconds.
+     */
+    int listen_timeout;
+#endif
+
+    /**
+     * Whether the cursor should be rendered on the server (remote) or on the
+     * client (local).
+     */
+    int remote_cursor;
```
Comment (jmuehlner) [inline src/protocols/vnc/settings.h]: No bool here?

File: src/protocols/vnc/vnc.c [hunk 4]
```diff
@@ -0,0 +1,404 @@
...
+
+    guac_socket_flush(client->socket);
+
+    guac_timestamp last_frame_end = guac_timestamp_current();
+
+    /* Handle messages from VNC server while client is running */
+    while (client->state == GUAC_CLIENT_RUNNING) {
+
+        /* Wait for start of frame */
+        int wait_result = guac_vnc_wait_for_messages(rfb_client, 1000000);
```
Comment (jmuehlner) [inline src/protocols/vnc/vnc.c]: What's this magic number 1000000?

File: src/protocols/vnc/settings.h (hunk 3 above)
Comment (mike-jumper) [inline src/protocols/vnc/settings.h]: It should be, yeah.

File: src/protocols/vnc/vnc.c (hunk 4 above)
Comment (mike-jumper) [inline src/protocols/vnc/vnc.c]: Uhhhhhhhh...

Well, I'll find out and fix that, too, I guess.

File: src/protocols/vnc/display.c (hunk 2 above)
Comment (jmuehlner) [inline src/protocols/vnc/display.c]: Ok

**Pull request merged**